hack: name for target ARM architecture not specified

crazy-max opened the pull request:
**- What I did**

Build currently doesn't set the right name for target ARM architecture through switches in `CGO_CFLAGS` and `CGO_CXXFLAGS` when doing cross-compilation. This was previously fixed in https://github.com/moby/moby/pull/43474

Also following changes for cross-compilation in https://github.com/moby/moby/pull/44546, we forgot to remove the toolchain configuration that is not used anymore as `xx` already sets correct cc/cxx envs already.

I was also looking at `go env` output while debugging and I found out `make.sh` overwrites `PKG_CONFIG` if not defined and set it to `pkg-config`. This is "problematic" when a build is invoked using `xx` in our Dockerfile because it already sets `PKG_CONFIG` to the right value in go environments depending on the target architecture: https://github.com/tonistiigi/xx/blob/8015613cccc2d3b689bf8189ec3604c4e4c94c34/base/xx-go#L75-L78. Last commit ensures we set the right value in our Dockerfile.

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

neersighted commented on hack/make/.binary:
```diff
@@ -16,61 +16,88 @@ source "${MAKEDIR}/.go-autogen"
 	fi
 
+	iscross=0
+	sandboxed=0
```
It looks like "sandboxed" really means "is xx in play, overriding the default toolchain names with the cross toolchain, so we don't need to address it by name"? Could we make it more descriptive, e.g. `isxx`? Also, is the availability of `xx-info` enough to detect this? I don't think we should mutate behavior here because a command is in someone's PATH.

neersighted commented on hack/make/.binary:
```diff
@@ -16,61 +16,88 @@ source "${MAKEDIR}/.go-autogen"
+		iscross=1
+	fi
+
+	if [[ "$sandboxed" = "0" ]] && [[ "$iscross" = "1" ]]; then
```
I'd like a comment here -- this appears to be "select a cross toolchain using a triple of XX isn't doing it for us and we're cross-compiling"

crazy-max replied:
Yes that's the right assumption, will add a comment.

neersighted commented on hack/make/.binary:
```diff
@@ -16,61 +16,88 @@ source "${MAKEDIR}/.go-autogen"
+		esac
+	fi
+
+	if [ "$iscross" = "1" ]; then
```
Can we make this comment a little more useful? Something like "since 32-bit ARM shares a toolchain, we must set the `-march` flag to select the correct subarchitecture"

neersighted commented on Dockerfile:
```diff
@@ -551,6 +551,7 @@ ARG TARGETPLATFORM
 RUN --mount=type=cache,sharing=locked,id=moby-build-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-build-aptcache,target=/var/cache/apt \
         xx-apt-get install --no-install-recommends -y \
+            dpkg-dev \
```
Why do we now need `dpkg-dev`?

crazy-max replied:
https://github.com/moby/moby/pull/44812#issuecomment-1380699254

neersighted replied:
Ah, you pushed as I was reviewing.

neersighted commented on hack/make.sh:
```diff
@@ -26,7 +26,7 @@ set -o pipefail
 export SCRIPTDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 export MAKEDIR="$SCRIPTDIR/make"
-export PKG_CONFIG=${PKG_CONFIG:-pkg-config}
+export PKG_CONFIG=${PKG_CONFIG:-$(go env PKG_CONFIG)}
```
`PKG_CONFIG` is not a standard part of `go env` -- is the mechanism XX uses to set it really just setting it in the `goenv` file? Why doesn't a regular fall-through (xx exporting it) work here?

crazy-max replied:
It seems to: https://github.com/golang/go/blob/245e95dfabd77f337373bf2d6bb47cd353ad8d74/src/internal/cfg/cfg.go#L67

neersighted replied:
Hmm, by standard I guess I mean "set by default" -- I'm not entirely familiar with how things are invoked.

Looking a little further, it seems like xx is replacing the go compiler with a shell script wrapper -- so to get an environment variable from xx, we need to interrogate it instead of letting it bubble down through the process tree?

crazy-max committed:
hack: typo in displayed platform name

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max committed:
Dockerfile: fix PKG_CONFIG for xx

Current implementation in hack/make.sh overwrites PKG_CONFIG
if not defined and set it to pkg-config. When a build is invoked
using xx in our Dockerfile, it will set PKG_CONFIG to the right
value in go environments depending on the target architecture: https://github.com/tonistiigi/xx/blob/8015613cccc2d3b689bf8189ec3604c4e4c94c34/base/xx-go#L75-L78

Also needs to install dpkg-dev to use pkg-config when cross-building

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

crazy-max commented:
opensuse mirror issue again :disappointed:

https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-44812/3/pipeline#step-140-log-1203

```
[2023-01-16T10:05:48.777Z] #66 39.32   502  Bad Gateway [IP: 195.135.221.134 443]
[2023-01-16T10:05:48.777Z] #66 39.32 Fetched 1506 kB in 1s (1724 kB/s)
[2023-01-16T10:05:48.777Z] #66 39.34 E: Failed to fetch [https://download.opensuse.org/repositories/devel:/tools:/criu/Debian_11/./arm64/criu_3.17.1-1_arm64.deb](https://download.opensuse.org/repositories/devel:/tools:/criu/Debian_11/arm64/criu_3.17.1-1_arm64.deb)  502  Bad Gateway [IP: 195.135.221.134 443]
[2023-01-16T10:05:48.777Z] #66 39.34 E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

neersighted commented on Dockerfile:
```diff
@@ -579,6 +580,7 @@ RUN --mount=type=bind,target=. \
   set -e
   target=$([ "$DOCKER_STATIC" = "1" ] && echo "binary" || echo "dynbinary")
   xx-go --wrap
+  export PKG_CONFIG=$(xx-go env PKG_CONFIG)
```
This still feels a little hacky to me; I feel like prefixing `PKG_CONFIG=` to just `make.sh` might be a bit more clear as to intent.

neersighted commented on hack/make/.binary:
```diff
@@ -16,60 +16,25 @@ source "${MAKEDIR}/.go-autogen"
-				linux/arm64)
-					export CC="${CC:-aarch64-linux-gnu-gcc}"
-					export CGO_ENABLED=1
+	case "$(go env GOOS)/$(go env GOARCH)" in
```
Since we have a single case, just make this `if [ ]` instead.

neersighted commented on hack/make/.binary:
```diff
@@ -18,7 +18,8 @@ source "${MAKEDIR}/.go-autogen"
 		linux/arm)
-			case "${GOARM}" in
+			# specify name of the target ARM architecture for cross-compilation
+			case "$(go env GOARM)" in
```
Please squash this into the previous commit as it was an oversight introduced in the previous commit.

neersighted commented on Dockerfile:
```diff
@@ -551,6 +551,7 @@ ARG TARGETPLATFORM
 RUN --mount=type=cache,sharing=locked,id=moby-build-aptlib,target=/var/lib/apt \
     --mount=type=cache,sharing=locked,id=moby-build-aptcache,target=/var/cache/apt \
         xx-apt-get install --no-install-recommends -y \
+            dpkg-dev \
```
Let's mention this change in the commit message as well.

neersighted reviewed (changes requested):
Overall, way cleaner and better thought out. I'd like some minor tweaks to this patch series, but otherwise LGTM. I do think we should endeavor to keep any knowledge of `xx` out of the `hack/` folder (confine it to docker-bake/the Dockerfile), and this is a great step in that direction.

crazy-max committed:
hack: name for target ARM architecture not specified

Build currently doesn't set the right name for target ARM
architecture through switches in CGO_CFLAGS and CGO_CXXFLAGS
when doing cross-compilation. This was previously fixed in https://github.com/moby/moby/pull/43474

Also removes the toolchain configuration. Following changes for
cross-compilation in https://github.com/moby/moby/pull/44546,
we forgot to remove the toolchain configuration that is
not used anymore as xx already sets correct cc/cxx envs already.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

neersighted reviewed (approved):
LGTM. I wonder if we should set `-march` for every `GOOS`, as GCC is the only supported Cgo toolchain, but that is also very much a bridge we can cross if it ever becomes relevant.
